scionlab/models/core.py

```diff
@@ -734,6 +740,10 @@ class Interface(models.Model):
             overrides the Host's default bind IP.""")
     bind_port = models.PositiveIntegerField(null=True, blank=True)
 
+    # Imported here to avoid circular dependencies
+    from scionlab.models.vpn import VPNClient
+    vpn_client = models.ForeignKey(VPNClient, on_delete=models.SET_NULL, null=True)
```

Comment:
The main reason of this relation is to keep track to which _attachment_ the `VPNClient` belongs to.

The reason I added it here is because the `VPNClient` is gathered starting from a `Link`(which is now the entity that ultimately represents a _connection_ between a `UserAS` and an `AttachmentPoint`), meaning that it can be retrieved with a single database hit.
The same could be achieved as well by placing a reference to the `VPNClient` in the `Link`, but it doesn't look too clean to me, since it's a field that wouldn't make sense for all the other `Link`s not involving a `UserAS`.

On the other hand, placing a reference toward the `Interface` from the `VPNClient` would make things cleaner (and it's the way I first implemented it I think), but it would also mean that for the vast majority of the time we need to retrieve a `VPNClient` (always given that we start from a `Link`), we first need to get the `Interface` `pk`, and then we can retrieve the `VPNClient` associated with it, thus we hit the database twice.

Reply:
I understand. From a `VPNClient` object, you could get the corresponding `Interface` by searching for the interfaces corresponding to the related attachment point with matching IP address, e.g.:
```
links = Link.objects.filter(interfaceA__host=vpn_client.vpn.server, 
                            interfaceB__host=vpn_client.host,  
                            interfaceB__public_ip=vpn_client.ip)
ifaces = [link.interfaceB for link in links]
```

Note that this dance is only necessary if there are:
- multiple interfaces to the same attachment point
- only a subset of these interfaces use VPN

otherwise you can just get _the_ interface(s) related to this attachment point.


>  The same could be achieved as well by placing a reference to the VPNClient in the Link, but it doesn't look too clean to me, since it's a field that wouldn't make sense for all the other Links not involving a UserAS.

`Interface` is better than `Link`, but I think the same reasoning still applies (it doesn't make sense for Interfaces not involving a UserAS).


Maybe you've already taken this into account but I think should point this out here: if there are multiple links to the same VPN Server (i.e. the same Attachment Point), these MUST share the same `VPNClient` -- the clients are   running in the same subnet and the routing wouldn't work if they're multiple clients in the same subnet.
This seems to rule out having a relation from `VPNClient` to `Interface`.

Reply:
Well, if there can only be one `VPNClient` per `UserAS - AttachmentPoint` relation (which totally make sense), then this field is not needed at all. We just need to keep track of whether or not a VPN connection is used for that connection, which can now become a simple, less invasive, boolean field (which can be stored in the `Interface` meaning that in that end of a certain `Link` the connection is established through a VPN tunnel)